src/lake/Lake/Build/Module.lean

```diff
@@ -881,9 +888,53 @@ def Module.initFacetConfigs : DNameMap ModuleFacetConfig :=
 @[inherit_doc Module.initFacetConfigs]
 abbrev initModuleFacetConfigs := Module.initFacetConfigs
 
-/-!
-Definitions to support `lake setup-file` builds.
+/-! ## Top-Level Builds
+Definitions to support `lake setup-file` and `lake lean` builds.
+-/
+
+/--
+Computes the module setup of a workspace module being edited by the Lean language server,
+building its imports and other dependencies. Used by `lake setup-file`.
+
+Due to its exclusive use as a top-level build, it does not construct a proper trace state.
 -/
+private def setupEditedModule
+  (mod : Module) (header : ModuleHeader)
+: FetchM (Job ModuleSetup) := do
+  let extraDepJob ← mod.lib.extraDep.fetch
+  let directImports ← header.imports.mapM fun imp => do
+    return ⟨imp, ← findModule? imp.module⟩
+  let fileName := mod.relLeanFile.toString
+  let localImports := directImports.filterMap (·.module?)
+  let impInfoJob ← fetchImportInfo fileName mod.name header
+  let precompileImports ←
+    if mod.shouldPrecompile then
+      (← computeTransImportsAux fileName localImports).await
+    else
+      (← computePrecompileImportsAux fileName localImports).await
+  let impLibsJob ← Job.collectArray (traceCaption := "import dynlibs") <$>
+    mod.fetchImportLibs precompileImports mod.shouldPrecompile
+  let externLibsJob ← Job.collectArray (traceCaption := "package external libraries") <$>
+    if mod.shouldPrecompile then mod.pkg.externLibs.mapM (·.dynlib.fetch) else pure #[]
+  let dynlibsJob ← mod.dynlibs.fetchIn mod.pkg "module dynlibs"
+  let pluginsJob ← mod.plugins.fetchIn mod.pkg "module plugins"
+  extraDepJob.bindM (sync := true) fun _ => do
+  impInfoJob.bindM (sync := true) fun info => do
+  impLibsJob.bindM (sync := true) fun impLibs => do
+  externLibsJob.bindM (sync := true) fun externLibs => do
+  dynlibsJob.bindM (sync := true) fun dynlibs => do
+  pluginsJob.mapM fun plugins => do
+    let {dynlibs, plugins} ← computeModuleDeps impLibs externLibs dynlibs plugins
+    let transImpArts ← fetchTransImportArts directImports info.directArts !header.isModule
+    return {
+      name := mod.name
+      isModule := header.isModule
+      imports? := none
```

Comment:
This isn't important (it doesn't matter in the language server because of the fallback logic, after all), but why do we return `none` here even when `lake setup-file` was provided with a set of direct imports? Also, if `lake setup-file` isn't being provided with a module header (which doesn't happen in the language server control flow AFAICT), shouldn't `setup-file` provide those that it parsed from disc?

Reply:
As to your first question, `none` is used to tell Lean to use the parsed imports from the header. This would not be `none` if Lake preprocessed and modified the set of imports in any way. While Lake could always pass the full set of imports, this reduces the complexity of the JSON for the common case (where no preprocessing is done) .As for second, `lake setup-file` is provided the module header (as JSON via stdin).